sympde/topology/space.py

```
@@ -122,25 +122,27 @@ def __new__(cls, name, domain, shape, kind):
         if not isinstance(domain, BasicDomain):
             raise TypeError('> Expecting a BasicDomain object for domain')
 
-        obj = Basic.__new__(cls, name, domain)
-        obj._shape = shape
+        obj = Basic.__new__(cls)
+        obj._name   = name
+        obj._domain = domain
+        obj._shape  = shape
 
         # ...
         if kind is None:
             kind = 'undefined'
 
         assert(isinstance(kind, (str, SpaceType)))
 
-        kind_str = kind.lower()
-        assert(kind_str in ['h1', 'hcurl', 'hdiv', 'l2', 'undefined'])
-
-        kind = dtype_space_registry[kind_str]
+        if isinstance(kind, str):
+            kind_str = kind.lower()
+            assert(kind_str in ['h1', 'hcurl', 'hdiv', 'l2', 'undefined'])
 
-        if not isinstance(kind, SpaceType):
+            kind = dtype_space_registry[kind_str]
+        elif not isinstance(kind, SpaceType):
             raise TypeError('Expecting kind to be of SpaceType')
```

I think the code cannot enter this branch, given the `assert` statement on line 134. If you want to raise a nice error you should remove the assert statement and explicitly check the type of `kind`.

---

ah yes you are right